pkg/soc/soc.go

```diff
@@ -0,0 +1,243 @@
+	}
+	fmt.Fprintf(os.Stderr, "%x", ownerAddressBytes)
+	ownerAddress, err := NewOwner(ownerAddressBytes)
+	if err != nil {
+		return err
+	}
+	s.signer = signer
+	s.owner = ownerAddress
+	return nil
+}
+
+// OwnerAddress returns the ethereum address of the signer of the Update.
+func (s *Update) OwnerAddress() []byte {
+	return s.owner.address
+}
+
+// Address returns the soc Chunk address of the update.
+func (s *Update) Address() (swarm.Address, error) {
+	var err error
+	hasher := swarm.NewHasher()
+	_, err = hasher.Write(s.id)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	_, err = hasher.Write(s.owner.address)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	addressBytes := hasher.Sum(nil)
+	return swarm.NewAddress(addressBytes), nil
+}
+
+// UpdateFromChunk recreates an Update from swarm.Chunk data.
+func UpdateFromChunk(ch swarm.Chunk) (*Update, error) {
+	chunkData := ch.Data()
+	if len(chunkData) < minUpdateSize {
+		return nil, errors.New("less than minimum length")
+	}
+
+	// add all the data fields
+	update := &Update{}
+	cursor := 0
+
+	update.id = chunkData[cursor : cursor+IdSize]
+	cursor += IdSize
+
+	update.signature = chunkData[cursor : cursor+SignatureSize]
+	cursor += SignatureSize
+
+	spanBytes := chunkData[cursor : cursor+swarm.SpanSize]
+	span := binary.LittleEndian.Uint64(spanBytes)
+	update.span = int64(span)
+	cursor += swarm.SpanSize
+
+	update.payload = chunkData[cursor:]
+
+	bmtPool := bmtlegacy.NewTreePool(swarm.NewHasher, swarm.Branches, bmtlegacy.PoolSize)
+	bmtHasher := bmtlegacy.New(bmtPool)
+
+	// calculate the bmt hash of the update payload
+	err := bmtHasher.SetSpan(int64(span))
+	if err != nil {
+		return nil, err
+	}
+	_, err = bmtHasher.Write(update.payload)
+	if err != nil {
+		return nil, err
+	}
+	payloadSum := bmtHasher.Sum(nil)
+
+	// recover owner information
+	recoveredPublicKey, err := crypto.Recover(update.signature, payloadSum)
+	if err != nil {
+		return nil, err
+	}
+	recoveredEthereumAddress, err := crypto.NewEthereumAddress(*recoveredPublicKey)
+	if err != nil {
+		return nil, err
+	}
+	owner, err := NewOwner(recoveredEthereumAddress)
+	if err != nil {
+		return nil, err
+	}
+	update.owner = owner
+
+	return update, nil
+}
+
+// CreateChunk creates a new chunk with signed payload ready for submission to the swarm network
+// from the given update data.
+//
+// This method will fail if no signer has been defined.
+func (s *Update) CreateChunk() (swarm.Chunk, error) {
+	if s.signer == nil {
+		return nil, errors.New("signer missing")
+	}
+	bmtPool := bmtlegacy.NewTreePool(swarm.NewHasher, swarm.Branches, bmtlegacy.PoolSize)
+	bmtHasher := bmtlegacy.New(bmtPool)
+
+	// calculate the bmt hash of the update payload
+	err := bmtHasher.SetSpan(s.span)
+	if err != nil {
+		return nil, err
+	}
+	_, err = bmtHasher.Write(s.payload)
+	if err != nil {
+		return nil, err
+	}
+	payloadSum := bmtHasher.Sum(nil)
+
+	// sign the update
+	signature, err := s.signer.Sign(payloadSum)
```

Comment:
it is not the payloadSum that is signed. but the payloadSum against the id

Reply:
... as it looks like it could also be the address it's signed with. Also wonder why there is a dotted line.